[BUG] flink 1.13.1 exactly-once restore from checkpoint tips the transaction with this txdID not found

**Describe the bug**
```
2021-09-17 17:55:36
java.lang.RuntimeException: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException: java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException$TransactionNotFoundException: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(4,26)`not found 
	at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.recoverAndCommit(FlinkPulsarSinkBase.java:513)
	at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.recoverAndCommit(FlinkPulsarSinkBase.java:86)
	at org.apache.flink.streaming.api.functions.sink.TwoPhaseCommitSinkFunction.recoverAndCommitInternal(TwoPhaseCommitSinkFunction.java:414)
	at org.apache.flink.streaming.api.functions.sink.TwoPhaseCommitSinkFunction.initializeState(TwoPhaseCommitSinkFunction.java:364)
	at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.initializeState(FlinkPulsarSinkBase.java:247)
	at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSink.initializeState(FlinkPulsarSink.java:41)
	at org.apache.flink.streaming.util.functions.StreamingFunctionUtils.tryRestoreFunction(StreamingFunctionUtils.java:189)
	at org.apache.flink.streaming.util.functions.StreamingFunctionUtils.restoreFunctionState(StreamingFunctionUtils.java:171)
	at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.initializeState(AbstractUdfStreamOperator.java:96)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.initializeOperatorState(StreamOperatorStateHandler.java:118)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:290)
	at org.apache.flink.streaming.runtime.tasks.OperatorChain.initializeStateAndOpenOperators(OperatorChain.java:436)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreGates(StreamTask.java:582)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.call(StreamTaskActionExecutor.java:55)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.executeRestore(StreamTask.java:562)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runWithCleanUpOnFail(StreamTask.java:647)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restore(StreamTask.java:537)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:759)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:566)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.lang.NullPointerException
		at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.producerFlush(FlinkPulsarSinkBase.java:367)
		at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSink.producerFlush(FlinkPulsarSink.java:41)
		at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.producerClose(FlinkPulsarSinkBase.java:542)
		at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.close(FlinkPulsarSinkBase.java:308)
		at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSink.close(FlinkPulsarSink.java:41)
		at org.apache.flink.api.common.functions.util.FunctionUtils.closeFunction(FunctionUtils.java:41)
		at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.dispose(AbstractUdfStreamOperator.java:117)
		at org.apache.flink.streaming.runtime.tasks.StreamTask.disposeAllOperators(StreamTask.java:861)
		at org.apache.flink.streaming.runtime.tasks.StreamTask.runAndSuppressThrowable(StreamTask.java:840)
		at org.apache.flink.streaming.runtime.tasks.StreamTask.cleanUpInvoke(StreamTask.java:753)
		at org.apache.flink.streaming.runtime.tasks.StreamTask.runWithCleanUpOnFail(StreamTask.java:659)
		... 4 more
Caused by: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException: java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException$TransactionNotFoundException: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(4,26)`not found 
	at org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException.unwrap(TransactionCoordinatorClientException.java:130)
	at org.apache.pulsar.client.impl.transaction.TransactionCoordinatorClientImpl.commit(TransactionCoordinatorClientImpl.java:225)
	at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSinkBase.recoverAndCommit(FlinkPulsarSinkBase.java:502)
	... 19 more
Caused by: java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException$TransactionNotFoundException: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(4,26)`not found 
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at org.apache.pulsar.client.impl.transaction.TransactionCoordinatorClientImpl.commit(TransactionCoordinatorClientImpl.java:223)
	... 20 more
Caused by: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException$TransactionNotFoundException: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(4,26)`not found 
	at org.apache.pulsar.client.impl.TransactionMetaStoreHandler.getExceptionByServerError(TransactionMetaStoreHandler.java:372)
	at org.apache.pulsar.client.impl.TransactionMetaStoreHandler.handleEndTxnResponse(TransactionMetaStoreHandler.java:299)
	at org.apache.pulsar.client.impl.ClientCnx.handleEndTxnResponse(ClientCnx.java:965)
	at org.apache.pulsar.common.protocol.PulsarDecoder.channelRead(PulsarDecoder.java:401)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.apache.pulsar.shade.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at org.apache.pulsar.shade.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.apache.pulsar.shade.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.apache.pulsar.shade.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at org.apache.pulsar.shade.io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at org.apache.pulsar.shade.io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	at org.apache.pulsar.shade.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at org.apache.pulsar.shade.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at org.apache.pulsar.shade.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at org.apache.pulsar.shade.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more

```

**To Reproduce**
Steps to reproduce the behavior:
1. use pulsar sql connector, with options sample:
    ```
    'connector' ='pulsar',
    'topic' ='',
    'service-url' ='',
    'admin-url' ='',
    'scan.startup.mode' = 'earliest',
    'format' = 'avro',
    'properties.changelog.mode' = 'true',
    'sink.semantic' = 'exactly-once
    ```
2. write hudi table content to pulsar
3. canal job
4. insert new record to hudi table
5. rerun flink job and restore from lastest checkpoint

**Expected behavior**
flink job can restore from lastest checkpoint

broker log
```
2021-09-17T18:32:05.104849004+08:00 10:32:05.104 [pulsar-io-5-1] ERROR org.apache.pulsar.broker.service.ServerCnx - Send response error for end txn request.

2021-09-17T18:32:05.104871902+08:00 java.util.concurrent.CompletionException: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(3,27)`not found

2021-09-17T18:32:05.104876975+08:00 at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]

2021-09-17T18:32:05.104881054+08:00 at java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:743) [?:?]

2021-09-17T18:32:05.104884357+08:00 at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731) [?:?]

2021-09-17T18:32:05.104887659+08:00 at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2108) [?:?]

2021-09-17T18:32:05.104890664+08:00 at org.apache.pulsar.broker.TransactionMetadataStoreService.endTransaction(TransactionMetadataStoreService.java:248) ~[org.apache.pulsar-pulsar-broker-2.8.0.jar:2.8.0]

2021-09-17T18:32:05.104901512+08:00 at org.apache.pulsar.broker.service.ServerCnx.handleEndTxn(ServerCnx.java:1890) [org.apache.pulsar-pulsar-broker-2.8.0.jar:2.8.0]

2021-09-17T18:32:05.104905033+08:00 at org.apache.pulsar.common.protocol.PulsarDecoder.channelRead(PulsarDecoder.java:396) [org.apache.pulsar-pulsar-common-2.8.0.jar:2.8.0]

2021-09-17T18:32:05.104908602+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104911900+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104915464+08:00 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104918567+08:00 at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:200) [io.netty-netty-handler-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104921603+08:00 at io.netty.handler.flow.FlowControlHandler.channelRead(FlowControlHandler.java:162) [io.netty-netty-handler-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104924471+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104927589+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104933263+08:00 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104936921+08:00 at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324) [io.netty-netty-codec-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104941098+08:00 at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296) [io.netty-netty-codec-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104944067+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104946811+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104949678+08:00 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104952791+08:00 at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104956167+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104959259+08:00 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104964638+08:00 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [io.netty-netty-transport-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104967587+08:00 at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795) [io.netty-netty-transport-native-epoll-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104979403+08:00 at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480) [io.netty-netty-transport-native-epoll-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104985278+08:00 at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378) [io.netty-netty-transport-native-epoll-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104988230+08:00 at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [io.netty-netty-common-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104991391+08:00 at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [io.netty-netty-common-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104994503+08:00 at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty-netty-common-4.1.63.Final.jar:4.1.63.Final]

2021-09-17T18:32:05.104997555+08:00 at java.lang.Thread.run(Thread.java:829) [?:?]

2021-09-17T18:32:05.105000384+08:00 Caused by: org.apache.pulsar.transaction.coordinator.exceptions.CoordinatorException$TransactionNotFoundException: The transaction with this txdID `(3,27)`not found

2021-09-17T18:32:05.105008330+08:00 at org.apache.pulsar.transaction.coordinator.impl.MLTransactionMetadataStore.getTxnMeta(MLTransactionMetadataStore.java:193) ~[org.apache.pulsar-pulsar-transaction-coordinator-2.8.0.jar:2.8.0]

2021-09-17T18:32:05.105012074+08:00 at org.apache.pulsar.broker.TransactionMetadataStoreService.getTxnMeta(TransactionMetadataStoreService.java:207) ~[org.apache.pulsar-pulsar-broker-2.8.0.jar:2.8.0]

2021-09-17T18:32:05.105016594+08:00 ... 27 more
```

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSinkBase.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSinkBase.java
@@ -314,6 +314,7 @@ abstract class FlinkPulsarSinkBase<T>
         checkErroneous();
         producerClose();
         checkErroneous();
+        super.close();
     }
 
     protected Producer<T> getProducer(String topic) {
@@ -375,7 +376,7 @@ abstract class FlinkPulsarSinkBase<T>
             }
         }
 
-        if (transaction.isTransactional()) {
+        if (transaction != null && transaction.isTransactional()) {
             // we check the future was completed and add the messageId to list for persistence.
             List<CompletableFuture<MessageId>> futureList =
                     tid2FuturesMap.get(transaction.transactionalId);
@@ -538,6 +539,14 @@ abstract class FlinkPulsarSinkBase<T>
                 // abort the transaction again, then Pulsar will throw a duplicate operation error,
                 // we catch the error without doing anything to deal with it
                 log.debug("transaction {} is already committed...", transaction.transactionalId);
+            } catch (
+                    TransactionCoordinatorClientException.TransactionNotFoundException
+                            notFoundException) {
+                // In some cases, the transaction has been committed or aborted before the recovery,
+                // but Flink has not yet sensed it. When flink recover this job, it will commit or
+                // abort the transaction again, then Pulsar will throw a duplicate operation error,
+                // we catch the error without doing anything to deal with it
+                log.debug("transaction {} is not found...", transaction.transactionalId);
             } catch (TransactionCoordinatorClientException e) {
                 throw new IllegalStateException(e);
             }
@@ -564,6 +573,14 @@ abstract class FlinkPulsarSinkBase<T>
                 // abort the transaction again, then Pulsar will throw a duplicate operation error,
                 // we catch the error without doing anything to deal with it
                 log.debug("transaction {} is already aborted...", transaction.transactionalId);
+            } catch (
+                    TransactionCoordinatorClientException.TransactionNotFoundException
+                            notFoundException) {
+                // In some cases, the transaction has been committed or aborted before the recovery,
+                // but Flink has not yet sensed it. When flink recover this job, it will commit or
+                // abort the transaction again, then Pulsar will throw a duplicate operation error,
+                // we catch the error without doing anything to deal with it
+                log.debug("transaction {} is not found...", transaction.transactionalId);
             } catch (TransactionCoordinatorClientException e) {
                 throw new IllegalStateException(e);
             }

--- a/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarSinkExactlyOnceTest.java
+++ b/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarSinkExactlyOnceTest.java
@@ -1,0 +1,258 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.pulsar;
+
+import org.apache.flink.api.common.serialization.SimpleStringSchema;
+import org.apache.flink.api.common.typeutils.base.StringSerializer;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.streaming.api.operators.StreamSink;
+import org.apache.flink.streaming.connectors.pulsar.serialization.PulsarSerializationSchema;
+import org.apache.flink.streaming.connectors.pulsar.serialization.PulsarSerializationSchemaWrapper;
+import org.apache.flink.streaming.connectors.pulsar.table.PulsarSinkSemantic;
+import org.apache.flink.streaming.util.OneInputStreamOperatorTestHarness;
+import org.apache.flink.table.api.DataTypes;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.api.SubscriptionInitialPosition;
+import org.apache.pulsar.client.impl.conf.ClientConfigurationData;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/** test exactly-once for pulsar sink. */
+public class PulsarSinkExactlyOnceTest extends PulsarTestBaseWithFlink {
+
+    private TestFlinkPulsarSink<String> sinkFunction;
+    private String topic;
+    private OneInputStreamOperatorTestHarness<String, Object> harness;
+
+    @Before
+    public void setUp() throws Exception {
+        this.topic = newTopic();
+        this.setUpTestHarness();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        this.closeTestHarness();
+    }
+
+    private void setUpTestHarness() throws Exception {
+        final Properties properties = new Properties();
+        final PulsarSerializationSchemaWrapper<String> schemaWrapper =
+                new PulsarSerializationSchemaWrapper.Builder<>(new SimpleStringSchema())
+                        .useAtomicMode(DataTypes.STRING())
+                        .build();
+        this.sinkFunction =
+                new TestFlinkPulsarSink<String>(
+                        getAdminUrl(),
+                        Optional.of(topic),
+                        clientConfigurationData,
+                        properties,
+                        schemaWrapper,
+                        PulsarSinkSemantic.EXACTLY_ONCE);
+        this.harness =
+                new OneInputStreamOperatorTestHarness(
+                        new StreamSink(this.sinkFunction), StringSerializer.INSTANCE);
+        this.harness.setup();
+    }
+
+    private void closeTestHarness() throws Exception {
+        this.harness.close();
+    }
+
+    @Test
+    public void testSubsumedNotificationOfPreviousCheckpoint() throws Exception {
+        this.harness.open();
+        this.harness.processElement("42", 0L);
+        this.harness.snapshot(0L, 1L);
+        this.harness.processElement("43", 2L);
+        this.harness.snapshot(1L, 3L);
+        this.harness.processElement("44", 4L);
+        this.harness.snapshot(2L, 5L);
+        this.harness.notifyOfCompletedCheckpoint(2L);
+        this.harness.notifyOfCompletedCheckpoint(1L);
+        this.assertExactlyOnce(Arrays.asList("42", "43", "44"));
+    }
+
+    @Test
+    public void testNotifyOfCompletedCheckpoint() throws Exception {
+        this.harness.open();
+        this.harness.processElement("42", 0L);
+        this.harness.snapshot(0L, 1L);
+        this.harness.processElement("43", 2L);
+        this.harness.snapshot(1L, 3L);
+        this.harness.processElement("44", 4L);
+        this.harness.snapshot(2L, 5L);
+        this.harness.notifyOfCompletedCheckpoint(1L);
+        this.assertExactlyOnce(Arrays.asList("42", "43"));
+    }
+
+    @Test
+    public void testRestoreCheckpoint() throws Exception {
+        this.harness.open();
+        this.harness.processElement("42", 0L);
+        this.harness.snapshot(0L, 1L);
+        this.harness.processElement("43", 2L);
+        final OperatorSubtaskState snapshot = this.harness.snapshot(1L, 3L);
+        this.harness.notifyOfCompletedCheckpoint(1L);
+
+        int count = 100;
+        for (int i = 3; i < count; i++) {
+            this.harness.processElement(Integer.toString(41 + i), i);
+            this.harness.snapshot(i, i);
+            this.harness.notifyOfCompletedCheckpoint(i);
+        }
+        this.closeTestHarness();
+        this.setUpTestHarness();
+        this.harness.initializeState(snapshot);
+        this.assertExactlyOnce(Arrays.asList("42", "43"));
+    }
+
+    @Test
+    public void testFailBeforeNotify() throws Exception {
+        this.harness.open();
+        this.harness.processElement("42", 0L);
+        this.harness.snapshot(0L, 1L);
+        this.harness.processElement("43", 2L);
+        OperatorSubtaskState snapshot = this.harness.snapshot(1L, 3L);
+        this.sinkFunction.setWritable(false);
+
+        try {
+            this.harness.processElement("44", 4L);
+            this.harness.snapshot(2L, 5L);
+            fail("something should fail");
+        } catch (NotWritableException ignore) {
+        }
+
+        this.closeTestHarness();
+        this.sinkFunction.setWritable(true);
+        this.setUpTestHarness();
+        this.harness.initializeState(snapshot);
+        this.assertExactlyOnce(Arrays.asList("42", "43"));
+        this.closeTestHarness();
+    }
+
+    private void assertExactlyOnce(List<String> expectedValues) throws Exception {
+
+        final List<String> actualValues =
+                consumeMessage(topic, Schema.STRING, expectedValues.size(), 2000);
+
+        Collections.sort(actualValues);
+        Collections.sort(expectedValues);
+        assertEquals(expectedValues, actualValues);
+    }
+
+    public <T> List<T> consumeMessage(String topic, Schema<T> schema, int count, int timeout)
+            throws TimeoutException, ExecutionException, InterruptedException {
+        return CompletableFuture.supplyAsync(
+                        () -> {
+                            Consumer<T> consumer = null;
+                            try {
+                                consumer =
+                                        getPulsarClient()
+                                                .newConsumer(schema)
+                                                .topic(topic)
+                                                .subscriptionInitialPosition(
+                                                        SubscriptionInitialPosition.Earliest)
+                                                .subscriptionName("test")
+                                                .subscribe();
+                                List<T> result = new ArrayList<>(count);
+                                for (int i = 0; i < count; i++) {
+                                    final Message<T> message = consumer.receive();
+                                    result.add(message.getValue());
+                                    consumer.acknowledge(message);
+                                }
+                                consumer.close();
+                                return result;
+                            } catch (Exception e) {
+                                sneakyThrow(e);
+                                return null;
+                            } finally {
+                                IOUtils.closeQuietly(consumer, i -> {});
+                            }
+                        })
+                .get(timeout, TimeUnit.MILLISECONDS);
+    }
+
+    /** javac hack for unchecking the checked exception. */
+    @SuppressWarnings("unchecked")
+    public static <T extends Exception> void sneakyThrow(Exception t) throws T {
+        throw (T) t;
+    }
+
+    /**
+     * mock Pulsar Sink,Support for throwing unwritable exceptions.
+     *
+     * @param <T> record
+     */
+    public static class TestFlinkPulsarSink<T> extends FlinkPulsarSink<T> {
+
+        private boolean writable = true;
+
+        public TestFlinkPulsarSink(
+                String adminUrl,
+                Optional<String> defaultTopicName,
+                ClientConfigurationData clientConf,
+                Properties properties,
+                PulsarSerializationSchema serializationSchema,
+                PulsarSinkSemantic semantic) {
+            super(
+                    adminUrl,
+                    defaultTopicName,
+                    clientConf,
+                    properties,
+                    serializationSchema,
+                    semantic);
+        }
+
+        @Override
+        public void invoke(PulsarTransactionState<T> transactionState, T value, Context context)
+                throws Exception {
+            if (!writable) {
+                throw new NotWritableException("TestFlinkPulsarSink");
+            }
+            super.invoke(transactionState, value, context);
+        }
+
+        public void setWritable(boolean writable) {
+            this.writable = writable;
+        }
+    }
+
+    /** not writable exception. */
+    public static class NotWritableException extends RuntimeException {
+        public NotWritableException(String name) {
+            super(String.format("Pulsar [%s] is not writable", name));
+        }
+    }
+}

--- a/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarSinkExactlyOnceTest.java
+++ b/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarSinkExactlyOnceTest.java
@@ -101,7 +101,7 @@ public class PulsarSinkExactlyOnceTest extends PulsarTestBaseWithFlink {
         this.harness.snapshot(2L, 5L);
         this.harness.notifyOfCompletedCheckpoint(2L);
         this.harness.notifyOfCompletedCheckpoint(1L);
-        this.assertExactlyOnce(Arrays.asList("42", "43", "44"));
+        assertEquals(Arrays.asList("42", "43", "44"), getActualValues(3));
     }
 
     @Test
@@ -114,7 +114,7 @@ public class PulsarSinkExactlyOnceTest extends PulsarTestBaseWithFlink {
         this.harness.processElement("44", 4L);
         this.harness.snapshot(2L, 5L);
         this.harness.notifyOfCompletedCheckpoint(1L);
-        this.assertExactlyOnce(Arrays.asList("42", "43"));
+        assertEquals(Arrays.asList("42", "43"), getActualValues(2));
     }
 
     @Test
@@ -135,7 +135,7 @@ public class PulsarSinkExactlyOnceTest extends PulsarTestBaseWithFlink {
         this.closeTestHarness();
         this.setUpTestHarness();
         this.harness.initializeState(snapshot);
-        this.assertExactlyOnce(Arrays.asList("42", "43"));
+        assertEquals(Arrays.asList("42", "43"), getActualValues(2));
     }
 
     @Test
@@ -158,18 +158,14 @@ public class PulsarSinkExactlyOnceTest extends PulsarTestBaseWithFlink {
         this.sinkFunction.setWritable(true);
         this.setUpTestHarness();
         this.harness.initializeState(snapshot);
-        this.assertExactlyOnce(Arrays.asList("42", "43"));
+        assertEquals(Arrays.asList("42", "43"), getActualValues(2));
         this.closeTestHarness();
     }
 
-    private void assertExactlyOnce(List<String> expectedValues) throws Exception {
-
-        final List<String> actualValues =
-                consumeMessage(topic, Schema.STRING, expectedValues.size(), 2000);
-
+    private List<String> getActualValues(int expectedSize) throws Exception {
+        final List<String> actualValues = consumeMessage(topic, Schema.STRING, expectedSize, 2000);
         Collections.sort(actualValues);
-        Collections.sort(expectedValues);
-        assertEquals(expectedValues, actualValues);
+        return actualValues;
     }
 
     public <T> List<T> consumeMessage(String topic, Schema<T> schema, int count, int timeout)

--- a/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarTestBase.java
+++ b/pulsar-flink-connector/src/test/java/org/apache/flink/streaming/connectors/pulsar/PulsarTestBase.java
@@ -47,6 +47,7 @@ import org.apache.pulsar.shade.org.apache.commons.lang3.RandomStringUtils;
 import org.apache.pulsar.shade.org.apache.commons.lang3.StringUtils;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
+import org.testcontainers.containers.BindMode;
 import org.testcontainers.containers.PulsarContainer;
 import org.testcontainers.containers.output.Slf4jLogConsumer;
 import org.testcontainers.containers.wait.strategy.HttpWaitStrategy;
@@ -112,6 +113,10 @@ public abstract class PulsarTestBase extends TestLogger {
                     DockerImageName.parse(pulsarImage)
                             .asCompatibleSubstituteFor("apachepulsar/pulsar");
             pulsarService = new PulsarContainer(pulsar);
+            pulsarService.withClasspathResourceMapping(
+                    "pulsar/txnStandalone.conf",
+                    "/pulsar/conf/standalone.conf",
+                    BindMode.READ_ONLY);
             pulsarService.addExposedPort(2181);
             pulsarService.waitingFor(
                     new HttpWaitStrategy()

